src/Adapter/ContainerBuilder.php

```diff
@@ -47,26 +91,183 @@ class ContainerBuilder
+        $container->setParameter('kernel.debug', $this->isDebug);
+        $container->setParameter('kernel.environment', $this->environment);
+        $container->setParameter('kernel.cache_dir', _PS_CACHE_DIR_);
+    }
+
+    /**
+     * @param SfContainerBuilder $container
+     *
+     * @throws \Doctrine\DBAL\DBALException
+     */
+    private function initDoctrine(SfContainerBuilder $container)
+    {
+        $configFile = _PS_ROOT_DIR_ . '/app/config/config.php';
+        if (!file_exists($configFile)) {
+            return;
+        }
+        $moduleRepository = ModuleRepositoryFactory::getInstance()->getRepository();
+        if (null === $moduleRepository) {
+            return;
+        }
+        $config = require $configFile;
+        $activeModules = $moduleRepository->getActiveModules();
+
+        $container->registerExtension(new DoctrineExtension());
+        $container->loadFromExtension('doctrine', $config['doctrine']);
+
+        $doctrinePassFactory = new ModulesDoctrinePassListBuilder($activeModules);
+        $compilerPassList = $doctrinePassFactory->getCompilerPassList($activeModules);
+        /** @var CompilerPassInterface $compilerPass */
+        foreach ($compilerPassList as $compilerResourcePath => $compilerPass) {
+            $container->addCompilerPass($compilerPass);
+            if (is_dir($compilerResourcePath)) {
+                $container->addResource(new DirectoryResource($compilerResourcePath));
+            } elseif (is_file($compilerResourcePath)) {
+                $container->addResource(new FileResource($compilerResourcePath));
+            }
+        }
+    }
+
+    /**
+     * @param SfContainerBuilder $container
+     *
+     * @throws \Exception
+     */
+    private function loadServices(SfContainerBuilder $container)
+    {
+        $loader = new YamlFileLoader($container, new FileLocator(__DIR__));
+        $servicesPath = _PS_CONFIG_DIR_ . sprintf('services/%s/services_%s.yml', $this->containerName, $this->environment);
+        if (file_exists($servicesPath)) {
+            $loader->load($servicesPath);
+        }
+    }
+
+    /**
+     * Loops through all active modules and automatically include their autoload (if present).
+     * Needs to be done as earlier as possible in application lifecycle.
+     *
+     * @throws \Doctrine\DBAL\DBALException
+     */
+    private function loadModulesAutoloader()
+    {
+        $moduleRepository = ModuleRepositoryFactory::getInstance()->getRepository();
+        if (null !== $moduleRepository) {
+            $activeModules = $moduleRepository->getActiveModules();
+            foreach ($activeModules as $module) {
+                $autoloader = _PS_ROOT_DIR_ . '/modules/' . $module . '/vendor/autoload.php';
+
+                if (file_exists($autoloader)) {
+                    include_once $autoloader;
+                }
+            }
+        }
+    }
+
+    /**
+     * @return string
+     */
+    private function getEnvironment()
```

Comment:
please, would you mind to extract this useful function into its own class? I know at least 3 places in PrestaShop where we could use it and probably fixe some weirds issues (like integration tests that sometimes write on production database).

Reply:
We wanted to perform this in another PR but fine.
Where should I put this new class? `EnvironmentInterface` in the core? And an implementation in `Adapter`?
Also should it use a static function? Or rather work as a service which can be injected but manages fallback with defined constants?

Reply:
done